### PR TITLE
More options for jni build.gradle

### PR DESCRIPTION
**Pull request by beeender**
* Add option for building core from source code
* Add option for building JNI with debug symbols
* Add option for building JNI without stripping symbols

### CODE REVIEW

**Comment by beeender** [comment]
@realm/java I need to do more testing on this change. But please help me to do a quick check first. The gradle doesn't play nice with me.

File: realm/realm-jni/build.gradle [hunk 1]
```diff
@@ -81,6 +89,34 @@ if (project.clang instanceof String) {
 if (project.forceDownloadCore instanceof String) {
     ext.forceDownloadCore = Boolean.parseBoolean(project.forceDownloadCore)
 }
+if (project.stripSymbols instanceof String) {
```
Comment (emanuelez) [inline realm/realm-jni/build.gradle]: I think all of these `instanceof` should be removed in favor of:

```groovy
ext.stripSymbols = project.stripSymbols as Boolean
```

What do you think @zaki50 ?

Comment (zaki50) [inline realm/realm-jni/build.gradle]: but  `"false" as Boolean` is `true`...

Comment (emanuelez) [inline realm/realm-jni/build.gradle]: You are right. I even found this tweet :) https://twitter.com/kohsukekawa/status/575671438809636865

Comment (emanuelez) [inline realm/realm-jni/build.gradle]: `"false".toBoolean()` maybe?

Comment (zaki50) [inline realm/realm-jni/build.gradle]: I think so.

Comment (zaki50) [inline realm/realm-jni/build.gradle]: but `project.stripSymbols` seems always boolean since we define `ext.stripSymbols`.
Is that true?

**Comment by beeender** [comment]
@realm/java

**Comment by beeender** [comment]
retest this please

**Comment by beeender** [comment]
retest this please

**Comment by cmelchior** [comment]
retest this please

File: realm/realm-jni/build.gradle [hunk 2]
```diff
@@ -75,11 +84,24 @@ buildscript {
...
-    ext.clang = Boolean.parseBoolean(project.clang)
+if (ext.debugBuild) {
+    // Debug build should never strip symbols
+    ext.stripSymbols = false
 }
-if (project.forceDownloadCore instanceof String) {
-    ext.forceDownloadCore = Boolean.parseBoolean(project.forceDownloadCore)
+if (ext.coreSourcePath) {
+    // Run the "sh build.sh get-version" to get the core version.
+    new ByteArrayOutputStream().withStream { os ->
```
Comment (nhachicha) [inline realm/realm-jni/build.gradle]: you can do something like this which is simple IMO
```groovy
ext.coreVersion =  "sh build.sh get-version".execute([], project.rootDir).text.trim()
```

File: realm/realm-jni/build.gradle [hunk 3]
```diff
@@ -100,6 +122,22 @@ def getNdk() {
     return ndkDir
 }
 
+def getStrippedExt() {
+    if (stripSymbols) {
```
Comment (nhachicha) [inline realm/realm-jni/build.gradle]:
```groovy
return stripSymbols ? "-stripped" :  ""
```
maybe more concise

File: realm/realm-jni/build.gradle [hunk 4]
```diff
@@ -100,6 +122,22 @@ def getNdk() {
...
+        return "-stripped"
+    } else {
+        return ""
+    }
+}
+
+def getDebugExt() {
+    if (debugBuild) {
```
Comment (nhachicha) [inline realm/realm-jni/build.gradle]: ditto

File: realm/realm-jni/build.gradle [hunk 5]
```diff
@@ -163,8 +201,44 @@ task downloadCore(group: 'build setup', description: 'Download the latest versio
...
+                    "build.sh",
+                    "build-android"
+            ]
+        }
+    }
+
+    // Copy the core tar ball
+    doLast {
+        println "${coreSourcePath}/realm-core-android-${coreVersion}.tar.gz"
+        println project.coreArchiveFile.getParent()
```
Comment (nhachicha) [inline realm/realm-jni/build.gradle]: *troll comment* 😈
you can reference the property without the getter 
```groovy
project.coreArchiveFile.parent
```

File: realm/realm-jni/build.gradle (hunk 2 above)
Comment (beeender) [inline realm/realm-jni/build.gradle]: No, it doesn't work... Did you miss anything? `"sh build.sh get-version"` is a string which doesn't have a `execute` method.

Comment (zaki50) [inline realm/realm-jni/build.gradle]: `ext.coreVersion =  "sh build.sh get-version".execute([], file(coreSourcePath)).text.trim()` works?

**Comment by beeender** [comment]
retest this please

**Comment by beeender** [comment]
retest this please

File: realm/realm-jni/build.gradle (hunk 2 above)
Comment (beeender) [inline realm/realm-jni/build.gradle]: Works! that is pure magic!

**Comment by beeender** [comment]
@realm/java Please help to review again.

File: realm/realm-jni/build.gradle [hunk 6]
```diff
@@ -75,11 +84,13 @@ buildscript {
...
-    ext.clang = Boolean.parseBoolean(project.clang)
+if (ext.debugBuild) {
+    // Debug build should never strip symbols
+    ext.stripSymbols = false
 }
-if (project.forceDownloadCore instanceof String) {
-    ext.forceDownloadCore = Boolean.parseBoolean(project.forceDownloadCore)
+if (ext.coreSourcePath) {
+    // Run the "sh build.sh get-version" to get the core version.
+    ext.coreVersion = "sh build.sh get-version".execute([], file(coreSourcePath)).text.trim()
```
Comment (cmelchior) [inline realm/realm-jni/build.gradle]: This won't work on Windows, but who cares :)

**Comment by cmelchior** [comment]
:+1:  But we should document these options. Maybe not in the README, but at least on the Wiki

**Comment by beeender** [comment]
Yes, I will document here. https://github.com/realm/realm-wiki/wiki/Debug-native-code

File: realm/realm-jni/build.gradle [hunk 7]
```diff
@@ -163,11 +182,40 @@ task downloadCore(group: 'build setup', description: 'Download the latest versio
     }
 }
 
+task compileCore(group: 'build setup', description: 'Compile the core library from source code') {
+    // Build the library from core source code
+    doFirst {
+        if (!coreSourcePath) {
+            throw new GradleException('The coreSourcePath is not set.')
+        }
+        exec {
```
Comment (zaki50) [inline realm/realm-jni/build.gradle]: This `exec` can be `'sh build.sh build-android'.execute([], file(coreSourcePath))`?

File: realm/realm-jni/build.gradle [hunk 8]
```diff
@@ -268,6 +316,7 @@ task clean(type: Delete) {
 
     delete fileTree(dir: "${projectDir}/../realm-library/src/main/jniLibs/", include: '**/librealm-jni*.so')
     delete fileTree(dir: "${projectDir}/../build/output/jniLibs-unstripped/", include: '**/librealm-jni*.so')
+    delete fileTree(dir: "${projectDir}/src/", include: '**/librealm-jni*-stripped.so')
```
Comment (zaki50) [inline realm/realm-jni/build.gradle]: Files under src should be cleaned by src/Makefile?

Comment (beeender) [inline realm/realm-jni/build.gradle]: this so is not created by Makefile (not directly) and `make clean` is called below.

File: realm/realm-jni/build.gradle (hunk 7 above)
Comment (beeender) [inline realm/realm-jni/build.gradle]: Tried it, ti doesn't run with `'sh build.sh build-android'.execute([], file(coreSourcePath))`. no idea why...

Comment (beeender) [inline realm/realm-jni/build.gradle]: And i suppose with `exec` block, the return code will be somehow caught by the `doFirst`?

File: realm/realm-jni/build.gradle (hunk 8 above)
Comment (zaki50) [inline realm/realm-jni/build.gradle]: Hmm, I understand that Makefile does not know `*-stripped.so`...

**Comment by zaki50** [comment]
:+1:

**Pull request merged**